src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -427,9 +427,20 @@ def __init__(
         self,
         base: BaseCandidate,
         extras: FrozenSet[str],
+        *,
+        comes_from: Optional[InstallRequirement] = None,
     ) -> None:
+        """
+        :param ireq: the InstallRequirement that led to this candidate if it
+            differs from the base's InstallRequirement. This will often be the
+            case in the sense that this candidate's requirement has the extras
+            while the base's does not. Unlike the InstallRequirement backed
+            candidates, this requirement is used solely for reporting purposes,
+            it does not do any leg work.
```

Comment:
This isn't true -- it is serving to allow modelling the extras' dependencies to the underlying package within the resolution context.

Reply:
As far as I can see this is [passed to `pip._internal.req.constructors.install_req_from_req_string`](https://github.com/pypa/pip/blob/0827d76bd2987ddea6f327e0c24e821d27d71cdf/src/pip/_internal/cli/req_command.py#L347), which only uses it for `comes_from` reporting. The `extras` parameter is the one used to model the extras' dependencies.

There is the unfortunate aspect that technically, `Resolver` can be instantiated with any function matching the [`InstallRequirementsProvider` signature](https://github.com/pypa/pip/blob/0827d76bd2987ddea6f327e0c24e821d27d71cdf/src/pip/_internal/resolution/base.py#L6), which does not declare explicitly what its arguments mean, or should be used for.

Am I missing something? If so, could you elaborate?

Reply:
Bah, I misread what was being stated here.